src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,173 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+
+#ifndef NUPIC_UTIL_SLIDING_WINDOW_HPP
+#define NUPIC_UTIL_SLIDING_WINDOW_HPP
+
+#include <vector>
+#include <algorithm>
+
+#include <nupic/types/Types.hpp>
+#include <nupic/utils/Log.hpp>
+
+namespace nupic {
+  namespace util {
+
+    template<class T> 
+    class SlidingWindow {
+      public:
+        SlidingWindow(UInt maxCapacity);
+        SlidingWindow(UInt maxCapacity, std::vector<T> initialData);
+        const UInt maxCapacity;
+        size_t size() const;
+        /** append new value to the end of the buffer and handle the "overflows"-may pop the first element if full. 
+        */
+        void append(T newValue);
+
+        /** like append, but return the dropped value. isValid indicates if the return value is valid (not while size()< maxCapacity)
+          :param T newValue - new value to append to the sliding window
+          :param bool isValid - a return pass-by-value that indicates validity of the return T value. for first maxCapacity items it is false, later always true.
+          :return T dropped value (past oldest element) if isValid; if not valid, this field holds the oldest value (but still contained in the window!)
+        */
+        T append(T newValue, bool& isValid);
+
+        /**
+        * :return unordered content (data ) of this sl. window; call getLinearizedData() if you need them oredered from oldest->newest
+        * This direct access method is fast.
+        */
+        const std::vector<T>& getData() const;
+
+        /** linearize method for the internal buffer; this is slower than the pure getData() but ensures that the data are ordered (oldest at the 
+         * beginning, newest at the end of the vector
+         * This handles case of |5,6;1,2,3,4| => |1,2,3,4,5,6|
+         * :return new linearized vector
+        */
+        std::vector<T> getLinearizedData() const;
+
+        bool operator==(const SlidingWindow& r2) const;
+        bool operator!=(const SlidingWindow& r2) const;
+        T& operator[](UInt index);
+        const T& operator[](UInt index) const;
+
+
+      private:
+        std::vector<T> buffer_;
+        UInt idxNext_;
+        size_t size_;
+}; 
+}} //end ns
+
+/// IMPLEMENTATION
+#include <cmath>
+
+using nupic::util::SlidingWindow;
+
+template<class T>
+SlidingWindow<T>::SlidingWindow(nupic::UInt maxCapacity) :
+  maxCapacity(maxCapacity) {
+  NTA_CHECK(maxCapacity > 0);
+  buffer_.reserve(maxCapacity);
+  buffer_.resize(maxCapacity);
```

Comment:
This requires the T class to have a default constructor. If it grows the vector, it initializes each T by calling T(). So the SlidingWindow is less reusable.

I'm not sure there's a way to solve this with the current approach. You could use a strategy that avoids the constructor (e.g. a C array + malloc), but that would cause bugs if you return a junk value and the destructor is called on the junk value.

You could solve this by going back to the previous approach which doesn't use `resize`.